### ISSUE TITLE
transaction.name not being captured in express in case of path parameter.

### ISSUE THREAD

**Issue by agarwalarjun123** [elastic/apm-agent-nodejs#2741]
**Describe the bug**

transaction.name does not include the base router url in express.

**To Reproduce**

1. Call API http://localhost:3000/v1/vital/:id
2. See error
{"log.level":"debug","@timestamp":"2022-05-31T12:46:51.832Z","log":{"logger":"elastic-apm-node"},"ecs":{"version":"1.6.0"},"message":"ended transaction {\"trans\":\"xxxxxxxxxxxx\",\"trace\":\"xxxxxxxxxxxxx\",\"type\":\"request\",\"result\":\"HTTP 2xx\",\"name\":\"GET /:id\"}"}

**Expected behavior**

transaction.name should come as /v1/vital/:id

**Environment**

- OS: Linux
- Node.js version: 14
- APM Server version: 7.17.3
- Agent version: 3.34.0

**How are you starting the agent? (please tick one of the boxes)**

- [X] Calling `agent.start()` directly (e.g. `require('elastic-apm-node').start(...)`)
- [ ] Requiring `elastic-apm-node/start` from within the source code
- [ ] Starting node with `-r elastic-apm-node/start`

**Additional context**
- Agent config options
  environment: config.NODE_ENV,
  serviceName: config.APM.SERVICE_NAME,
  serverUrl: _.get(config, 'APM.SERVICE_URI'),
  captureBody: 'all',
  captureHeaders: true,
  logUncaughtExceptions: true,
  errorOnAbortedRequests: true,
  captureErrorLogStackTraces: 'always',
  usePathAsTransactionName: true,
  logLevel: 'trace',

**Comment by trentm** [comment]
Hi @agarwalarjun123. Thanks for the issue.

Here is the code in the APM agent that is determining the default transaction name from the incoming HTTP request info:

```js
function getPathFromRequest (req, useBase, usePathAsTransactionName) {
  if (req[symbols.staticFile]) {
    return 'static file'
  }

  var path = getStackPath(req)
  var route = routePath(req.route)

  if (route) {
    return path ? join([path, route]) : route
  } else if (path && (path !== '/' || useBase)) {
    return path
  }

  if (usePathAsTransactionName) {
    const parsed = parseUrl(req)
    return parsed && parsed.pathname
  }
}
```

What web framework are you using? Express? or some other one?

Are you able to post the parts of your server code that setup the routes? That might help us figure out what the issue is. My guess is that the request `req` has `req.route.path` set to `/:id`. Perhaps that is because of how the `/v1/vital` route is mounted?

**Comment by agarwalarjun123** [comment]
yes, we are using express. We have 3 level routing for the express app.
**main app router**
app.use('/', router);

**versionRouter**
router.use('/v1',versionRouter)

**module router**
versionRouter.router('/vital', vitalRouter)

**vitalRouter**
vitalRouter.get('/:id', controller);

what I understand from here is that req.route.path gives us path relative to the vital router? (/:id). can we use req.baseUrl for getting the request baseUrl.

**Comment by trentm** [comment]
Thanks very much for that reference snippet. I can reproduce the issue. I haven't dug in, but I assume this is a bug in the APM agent.

> Also to note here, that in some cases it does work correctly.

Are there any differences in how those are mounted that you can think of? It might really help in fixing the issue.

**Comment by trentm** [comment]
The difference is whether the request handler defers to `next`, or if it handles it itself.

```js
// https://github.com/elastic/apm-agent-nodejs/issues/2741
require('./').start({ // elastic-apm-node
  serviceName: 'play-express-3-level-routes',
  captureBody: 'all',
  captureHeaders: true,
  metricsInterval: '0s',
  logLevel: 'debug'
})
const express = require('express')

const app = express()
const moduleRouter = express.Router()
moduleRouter.get('/:id', (req, res, next) => {
  return res.json({ data: true }) // `GET /v1/vital/:id` with this (correct)
  // return next({ data: true })  // `GET /:id` with this (wrong)
})

const vitalRouter = express.Router()
vitalRouter.use('/vital', moduleRouter)
app.use('/v1', vitalRouter)

app.use(function jsonIt (data, req, res, _next) {
  return res.json(data)
})

app.listen(3000, () => console.log('server running on port 3000'))
```

I haven't dug any further yet.

**Comment by trentm** [comment]
@agarwalarjun123 I no longer believe this is a bug in the agent.

https://expressjs.com/en/guide/error-handling.html says this:

> If you pass anything to the `next()` function (except the string `'route'`), Express regards the current request as being an error and will skip any remaining non-error handling routing and middleware functions.

and

> Define error-handling middleware functions in the same way as other middleware functions, except error-handling functions have four arguments instead of three

So the `app.use((data, req, res, _next) => res.json(data))` you defined above is an error-handling middleware function according to Express routing. And when you call `next({ data: true })` in the `moduleRouter.get('/:id'` handler, the request is considered to be in error and your error-handling middleware function gets called.

So this feels to me like accidental incorrect usage of Express routing.

The APM agent's instrumentation of Express routing *does* watch for an Error instance being passed to that `next(...)` callback and will result in the correct `transaction.name` in that case.

Looking at the express routing code, it actually special-cases the exact strings 'route' and 'router' and considers any other truthy value an error, so I think the APM agent's Express instrumentation could improve a bit here.

**Comment by trentm** [comment]
@agarwalarjun123 While I think your usage of Express routing in your example is not how Express was intended to be used, you helped me find a bug in the APM agent.  Express considers `next({ some: 'value'})` to be an *error* case and calls any registered error handlers. The APM agent currently didn't handle that properly. I have a fix in https://github.com/elastic/apm-agent-nodejs/pull/2750. Once that fix is in, then for your case the created APM transaction will be the `GET /v1/vital/:id` that you expect.

**Comment by trentm** [comment]
@agarwalarjun123 I hope to have a new release out early next week.
